src/main/kotlin/core/controller/UserController.kt

```diff
@@ -0,0 +1,49 @@
+package core.controller
+
+import core.model.User
+import core.model.UserRegistration
+import org.springframework.http.HttpStatus
+import org.springframework.web.bind.annotation.*
+import org.springframework.web.server.ResponseStatusException
+import core.service.UserService
+
+@RestController
+@RequestMapping("api/v1/user")
+class UserController(val userService: UserService) {
+    @PostMapping("/register")
+    fun registerUser(
+        @RequestBody registration: UserRegistration
+    ) {
+        val existing = userService.findUser(registration.username)
+
+        if (existing != null) {
+            throw ResponseStatusException(
+                HttpStatus.CONFLICT, "User already exists", null
+            )
+        }
+
+        val user = User(null, registration.username, registration.password)
+        userService.post(user)
```

Comment:
I would expect that the service function name would be more descriptive, similar to the `registerUser` that you used in the controller.